File: trieshake/reporter.py
"""Reporter module — output formatting and plan CSV export."""

import csv
from pathlib import Path, PurePosixPath

from trieshake.planner import PlanEntry


def print_scan_result(count: int) -> None:
    print(f"\U0001F50D Scanning — found {count} matching files.")


def print_plan_summary(plans: list[PlanEntry], dry_run: bool = True) -> None:
    moves = sum(
        1 for p in plans
        if str(p.source_path) != str(p.target_dir / p.target_filename)
    )
    skips = len(plans) - moves
    collisions = sum(1 for p in plans if p.is_collision)

    print(f"\U0001F4CB Building plan — {len(plans)} files total.")
    print(f"   {moves} to move, {skips} already correct.")
    if collisions:
        print(f"   \u26A0\uFE0F  {collisions} collisions detected.")

    if plans:
        p = plans[0]
        print(f"   Example: {p.source_path} -> {p.target_dir / p.target_filename}")

    if dry_run:
        print(f"\n\U0001F6AB Dry run — no files moved. Use --execute to apply.")


def print_move_result(
    successes: list[PlanEntry],
    failures: list[tuple[PlanEntry, str]],
) -> None:
    print(f"\U0001F4E6 Moving files — {len(successes)} succeeded, {len(failures)} failed.")
    for plan, err in failures:
        print(f"   \u274C {plan.source_path}: {err}")


def print_verify_result(lost: list[PlanEntry]) -> None:
    if lost:
        print(f"\u26A0\uFE0F  Verification — {len(lost)} files lost!")
        for p in lost:
            print(f"   Missing: {p.target_dir / p.target_filename}")
    else:
        print("\u2705 Verification — all files accounted for.")


def print_cleanup_result(removed: list[Path]) -> None:
    print(f"\U0001F9F9 Cleanup — removed {len(removed)} directories.")


def print_collision_report(plans: list[PlanEntry]) -> None:
    collisions = [p for p in plans if p.is_collision]
    if not collisions:
        return
    print(f"\n\U0001F4CB Collision report ({len(collisions)} collisions):")
    for p in collisions:
        print(f"   {p.source_path} -> {p.target_dir / p.target_filename}")
        print(f"      collides with: {p.collision_of}")


def export_plan_csv(plans: list[PlanEntry], output_path: Path) -> None:
    """Write the move plan to a CSV file."""
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow([
            "source_path", "target_path", "leafname", "concat_string",
            "prefix_groups", "is_collision", "collision_of",
            "already_correct", "action",
        ])
        for p in plans:
            target_path = str(p.target_dir / p.target_filename)
            already_correct = str(p.source_path) == target_path
            if p.is_collision:
                action = "collision"
            elif already_correct:
                action = "skip"
            else:
                action = "move"
            writer.writerow([
                str(p.source_path),
                target_path,
                p.leafname,
                p.concat_string,
                "|".join(p.chunks),
                str(p.is_collision).lower(),
                p.collision_of or "",
                str(already_correct).lower(),
                action,
            ])

File: clojure/src/trieshake/reporter.clj
(ns trieshake.reporter
  "Reporter module — output formatting and plan CSV export."
  (:require [clojure.string :as str]))

(defn print-scan-result [count]
  (println (str "\uD83D\uDD0D Scanning \u2014 found " count " matching files.")))

(defn print-plan-summary [plans dry-run?]
  (let [total (count plans)
        moves (count (filter (fn [p]
                               (not= (:source-path p)
                                     (str (:target-dir p) "/" (:target-filename p))))
                             plans))
        skips (- total moves)
        collisions (count (filter :is-collision plans))]
    (println (str "\uD83D\uDCCB Building plan \u2014 " total " files total."))
    (println (str "   " moves " to move, " skips " already correct."))
    (when (pos? collisions)
      (println (str "   \u26A0\uFE0F  " collisions " collisions detected.")))
    (when (seq plans)
      (let [p (first plans)]
        (println (str "   Example: " (:source-path p) " -> "
                      (:target-dir p) "/" (:target-filename p)))))
    (when dry-run?
      (println (str "\n\uD83D\uDEAB Dry run \u2014 no files moved. Use --execute to apply.")))))

(defn print-move-result [successes failures]
  (println (str "\uD83D\uDCE6 Moving files \u2014 " (count successes) " succeeded, "
                (count failures) " failed."))
  (doseq [[plan err] failures]
    (println (str "   \u274C " (:source-path plan) ": " err))))

(defn print-verify-result [lost]
  (if (seq lost)
    (do
      (println (str "\u26A0\uFE0F  Verification \u2014 " (count lost) " files lost!"))
      (doseq [p lost]
        (println (str "   Missing: " (:target-dir p) "/" (:target-filename p)))))
    (println "\u2705 Verification \u2014 all files accounted for.")))

(defn print-cleanup-result [removed]
  (println (str "\uD83E\uDDF9 Cleanup \u2014 removed " (count removed) " directories.")))

(defn print-collision-report [plans]
  (let [collisions (filter :is-collision plans)]
    (when (seq collisions)
      (println (str "\n\uD83D\uDCCB Collision report (" (count collisions) " collisions):"))
      (doseq [p collisions]
        (println (str "   " (:source-path p) " -> "
                      (:target-dir p) "/" (:target-filename p)))
        (println (str "      collides with: " (:collision-of p)))))))

(defn- csv-quote
  "Quote a field for CSV output if it contains commas, quotes, or newlines."
  [s]
  (let [s (str s)]
    (if (re-find #"[,\"\n\r]" s)
      (str "\"" (str/replace s "\"" "\"\"") "\"")
      s)))

(defn export-plan-csv
  "Write the move plan to a CSV file."
  [plans output-path]
  (let [header "source_path,target_path,leafname,concat_string,prefix_groups,is_collision,collision_of,already_correct,action"
        rows (map (fn [p]
                    (let [target-path (str (:target-dir p) "/" (:target-filename p))
                          already-correct (= (:source-path p) target-path)
                          action (cond
                                   (:is-collision p) "collision"
                                   already-correct "skip"
                                   :else "move")]
                      (str/join ","
                                (map csv-quote
                                     [(:source-path p)
                                      target-path
                                      (:leafname p)
                                      (:concat-string p)
                                      (str/join "|" (:chunks p))
                                      (str/lower-case (str (:is-collision p)))
                                      (or (:collision-of p) "")
                                      (str/lower-case (str already-correct))
                                      action]))))
                  plans)]
    (spit output-path (str header "\n" (str/join "\n" rows) "\n"))))
